File: Web/packages/kernel/test/watch-conformance.test.ts
//
//  watch-conformance.test.ts - the SHARED watch-dispatch corpus
//  (OpenSource/Conformance/actions/watch-dispatch.json) through the TS primitives a
//  surface wires: `store.watch` (fires on meaningful change, never on subscribe) feeding
//  the ActionRunner. The Kotlin (WatchConformanceTest.kt, the WatchView evaluate/fire
//  loop) and Swift (WatchConformance, ConformanceHosts.swift, record lane) twins run the
//  SAME file. Pinned here after the W12 stale-snapshot investigation: the handler
//  observes the POST-WRITE store (the filed 2026-08-17 starter toggle revert was the
//  wave-7 F4 entity lexing writing 0 through this dispatch path, never a snapshot).
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import { ReactiveStore, DSXState, flushEffects } from "../src/store.ts";
import { ActionRunner, makeRunEnv, writeBound } from "../src/runner.ts";
import { JSE } from "../src/jse/jse.ts";
import { NSNull, isDict, string, type Dict } from "../src/jse/values.ts";

function corpusFile(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/actions/watch-dispatch.json");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("watch-dispatch corpus not found");
    dir = parent;
  }
}

function toJse(v: unknown): unknown {
  if (v === null) return NSNull;
  if (Array.isArray(v)) return v.map(toJse);
  if (typeof v === "object") {
    const out: Dict = {};
    for (const [k, val] of Object.entries(v as Dict)) out[k] = toJse(val);
    return out;
  }
  return v;
}

type Case = {
  name: string;
  actions: { [name: string]: { inputs?: Dict; body: string } };
  scope: Dict;
  global?: Dict;
  watches: Array<{ value: string; handler: string }>;
  pre?: Array<{ path: string; value: unknown }>;
  run: string;
  runItem?: Dict;
  expectStore: { [path: string]: unknown };
  expectGlobal?: { [path: string]: unknown };
  expectEvents: string[];
};

const doc = JSON.parse(readFileSync(corpusFile(), "utf-8")) as { cases: Case[] };
assert.ok(doc.cases.length > 0, "watch-dispatch corpus is empty");

/** The reference payload rule (Stack.swift WatchView.fire / JseRunner.fireWatch):
 *  an object value rides dsx.this as-is, anything else as { value: … }. */
function payloadFor(v: unknown): Dict {
  return isDict(v) ? (v as Dict) : { value: v ?? NSNull };
}

for (const c of doc.cases) {
  test(`watch-corpus/${c.name}`, async () => {
    // a fresh app-wide store per case (DSXState is a module singleton)
    for (const k of Object.keys(DSXState.vars)) DSXState.set(k, null);
    for (const [k, v] of Object.entries(c.global ?? {})) DSXState.set(k, toJse(v));

    const store = new ReactiveStore();
    for (const [k, v] of Object.entries(c.scope)) store.jse.vars.set(k, toJse(v));

    const events: string[] = [];
    const env = makeRunEnv(store, { emitEvent: (name) => events.push(name) });
    for (const [name, decl] of Object.entries(c.actions ?? {})) {
      env.actions.set(name, { body: decl.body, inputs: (decl.inputs ?? {}) as Dict });
    }
    const runner = new ActionRunner(env);

    // watches attach exactly as a mounted surface attaches them: never fire on subscribe;
    // each fire dispatches one runner entry whose settling the drain below awaits.
    const inFlight: Array<Promise<void>> = [];
    const disposers = (c.watches ?? []).map((w) =>
      store.watch(
        () => store.eval(w.value, null),
        (v) => { inFlight.push(runner.run(w.handler, null, payloadFor(v))); },
      ));

    try {
      for (const p of c.pre ?? []) writeBound(env, p.path, toJse(p.value));
      inFlight.push(runner.run(c.run, c.runItem ? (toJse(c.runItem) as Dict) : null));

      // Drain the cascade: deliver scheduled watch evaluations, await every handler entry
      // they dispatched, repeat until a quiet round. Bounded - corpus rows are synchronous.
      for (let round = 0; round < 32; round++) {
        flushEffects();
        if (inFlight.length === 0) break;
        const batch = inFlight.splice(0, inFlight.length);
        await Promise.all(batch);
        await new Promise((r) => setTimeout(r, 0)); // entry-lock continuations land
      }
      flushEffects();

      const nil = (v: unknown): unknown => (v === null || v === undefined || v === NSNull ? null : v);
      for (const [path, expected] of Object.entries(c.expectStore)) {
        const actual = JSE.eval(path, store.jse, null);
        const expectVal = expected === null ? null : toJse(expected);
        assert.ok(
          JSE.equals(nil(actual), expectVal),
          `${path} -> ${string(actual)} (expected ${string(expected)})`,
        );
      }
      for (const [path, expected] of Object.entries(c.expectGlobal ?? {})) {
        const actual = DSXState.get(path);
        const expectVal = expected === null ? null : toJse(expected);
        assert.ok(
          JSE.equals(nil(actual), expectVal),
          `global.${path} -> ${string(actual)} (expected ${string(expected)})`,
        );
      }
      assert.deepEqual(events, c.expectEvents, "event order");
    } finally {
      for (const d of disposers) d();
      store.dispose();
    }
  });
}

File: Engine/Android/core/src/test/kotlin/despia/engine/WatchConformanceTest.kt
package despia.engine

import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import java.io.File
import java.util.concurrent.Executor
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The watch-dispatch conformance runner — executes
 * OpenSource/Conformance/actions/watch-dispatch.json through THIS runtime's dispatch funnel:
 * the WatchView evaluate loop (Compose StackNodeView.WatchView — evaluate `value`, compare
 * JSE.watchKey, fire on meaningful change, never on subscribe) feeding the REAL
 * `JSERunner.fireWatch` (payload rule + budget + afterRender dispatch). The TS harness
 * (watch-conformance.test.ts, store.watch → ActionRunner) and the Swift reference
 * (WatchConformance, ConformanceHosts.swift, record lane) run the SAME file.
 *
 * Pinned after the W12 stale-snapshot investigation: a watch handler observes the POST-WRITE
 * store — `run(change, payload, payload)` executes against the live shared store, so every
 * store read inside the handler sees the state that triggered the fire. (The filed 2026-08-17
 * starter toggle revert was the wave-7 F4 entity lexing assigning 0 through this path, never
 * a snapshot — jse/syntax-006 pins the decode, the corpus's entity-spelled row pins the hold.)
 *
 * The evaluate loop here is the :core stand-in for recomposition: after the bound-control
 * `pre` writes and after the entry, re-evaluate every watch and fire the changed ones until a
 * quiet round (bounded — corpus rows are synchronous, matching the natives' settled-fire
 * timing; fire COUNT within one multi-write entry stays unpinned by the corpus).
 */
class WatchConformanceTest {

    private fun corpusFile(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/actions/watch-dispatch.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/actions/watch-dispatch.json not found")
        }
    }

    private fun nil(v: Any?): Any? = if (v == NSNull) null else v

    @TestFactory
    fun watchCorpus(): List<DynamicTest> {
        val doc = json(corpusFile().readText()).foundationValue as? Map<*, *>
            ?: error("watch-dispatch.json: not a JSON object")
        @Suppress("UNCHECKED_CAST")
        val cases = doc["cases"] as? List<Map<String, Any?>> ?: error("watch-dispatch.json: no cases[]")
        return cases.map { c ->
            val name = "watch-corpus/${c["name"]}"
            DynamicTest.dynamicTest(name) {
                JSERunner.mainExecutor = Executor { it.run() }
                JSERunner.backgroundExecutor = Executor { it.run() }
                // The app-wide plane: DSX.state is a process singleton — bind the boot seam,
                // snapshot + restore its vars so `global.*` writes stay isolated per case.
                val savedStateVars = JSE.stateVars
                val savedDispatch = JSE.afterRenderDispatch
                val savedState = HashMap(DSX.state.vars)
                JSE.stateVars = { DSX.state.vars }
                JSE.afterRenderDispatch = { it() }
                try {
                    DSX.state.vars.clear()
                    @Suppress("UNCHECKED_CAST")
                    val globalSeed = (c["global"] as? Map<String, Any?>) ?: emptyMap()
                    globalSeed.forEach { (k, v) -> DSX.state.vars[k] = v }

                    val store = StackStore()
                    @Suppress("UNCHECKED_CAST")
                    val scope = (c["scope"] as? Map<String, Any?>) ?: emptyMap()
                    scope.forEach { (k, v) -> store.vars[k] = v }

                    @Suppress("UNCHECKED_CAST")
                    val actions = (c["actions"] as? Map<String, Map<String, Any?>>) ?: emptyMap()
                    for ((actionName, decl) in actions) {
                        @Suppress("UNCHECKED_CAST")
                        val inputs = ((decl["inputs"] as? Map<String, Any?>) ?: emptyMap())
                            .mapValues { JSE.string(it.value) }
                        store.registerAction(actionName, inputs, JSE.string(decl["body"]))
                    }

                    val events = ArrayList<String>()
                    store.anyHandlers.add { evName, _ -> events.add(evName) }

                    val runner = JSERunner(store)

                    // Subscribe: evaluate once, record the key, NEVER fire (no corpus row
                    // declares `immediate` — the WatchView prev==null arm).
                    @Suppress("UNCHECKED_CAST")
                    val watches = (c["watches"] as? List<Map<String, Any?>>) ?: emptyList()
                    val exprs = watches.map { JSE.string(it["value"]) }
                    val handlers = watches.map { JSE.string(it["handler"]) }
                    val last = exprs.map { JSE.watchKey(JSE.eval(it, store, null)) }.toMutableList()

                    // One recomposition settle: re-evaluate every watch, fire the meaningfully
                    // changed ones through the REAL dispatch funnel, repeat until quiet.
                    fun settle() {
                        for (round in 0 until 32) {
                            var fired = false
                            for (i in exprs.indices) {
                                val v = JSE.eval(exprs[i], store, null)
                                val key = JSE.watchKey(v)
                                if (key != last[i]) {
                                    last[i] = key
                                    fired = true
                                    runner.fireWatch(v, handlers[i], exprs[i])
                                }
                            }
                            if (!fired) break
                        }
                    }

                    @Suppress("UNCHECKED_CAST")
                    val pre = (c["pre"] as? List<Map<String, Any?>>) ?: emptyList()
                    for (p in pre) store.writeBound(JSE.string(p["path"]), p["value"] ?: NSNull)
                    settle()

                    @Suppress("UNCHECKED_CAST")
                    val runItem = c["runItem"] as? Map<String, Any?>
                    runner.run(JSE.string(c["run"]), runItem)
                    settle()

                    @Suppress("UNCHECKED_CAST")
                    val expectStore = (c["expectStore"] as? Map<String, Any?>) ?: emptyMap()
                    for ((path, expected) in expectStore) {
                        val actual = JSE.eval(path, store, null)
                        assertTrue(
                            JSE.equals(nil(actual), nil(expected)),
                            "$name: $path -> ${JSE.string(actual)} (expected ${JSE.string(expected)})",
                        )
                    }
                    @Suppress("UNCHECKED_CAST")
                    val expectGlobal = (c["expectGlobal"] as? Map<String, Any?>) ?: emptyMap()
                    for ((path, expected) in expectGlobal) {
                        val actual = DSX.state.getPath(path)
                        assertTrue(
                            JSE.equals(nil(actual), nil(expected)),
                            "$name: global.$path -> ${JSE.string(actual)} (expected ${JSE.string(expected)})",
                        )
                    }
                    @Suppress("UNCHECKED_CAST")
                    val expectEvents = ((c["expectEvents"] as? List<Any?>) ?: emptyList()).map { JSE.string(it) }
                    assertEquals(expectEvents, events, "$name: event order")
                } finally {
                    DSX.state.vars.clear()
                    DSX.state.vars.putAll(savedState)
                    JSE.stateVars = savedStateVars
                    JSE.afterRenderDispatch = savedDispatch
                }
            }
        }
    }
}
